### PR TITLE
Fix JFile::getExt() when path contains period and file has no extension

### PR DESCRIPTION
**Pull request by SharkyKZ**
### Summary of Changes

Fix JFile::getExt() when path contains period and file has no extension

### Testing Instructions

Pass filename without extension but with path containing a period to `JFile::getExt()`, e.g.

    JFile::getExt('C:/wamp/www/joomla.cms/somefile')

### Actual result BEFORE applying this Pull Request

>cms/somefile

### Expected result AFTER applying this Pull Request

Empty string.

### Documentation Changes Required

No.

### CODE REVIEW

File: libraries/src/Filesystem/File.php [hunk 1]
```diff
@@ -44,7 +44,15 @@ public static function getExt($file)
 			return '';
 		}
 
-		return (string) substr($file, $dot + 1);
+		$ext = substr($file, $dot + 1);
+
+		// Extension cannot contain slashes.
+		if (preg_match('#[/\\\\]#', $ext) === 1)
```
Comment (Fedik) [inline libraries/src/Filesystem/File.php]: maybe `strpos($ext, '/') !== false || strpos($ext, '\\') !== false` ?

Comment (SharkyKZ) [reply libraries/src/Filesystem/File.php]: Either way works. Not sure which one is faster.

Comment (HLeithner) [reply libraries/src/Filesystem/File.php]: If you have to start the regex engine you have a big penalty. So I would suggest to use the `strpos` instead.

Comment (SharkyKZ) [reply libraries/src/Filesystem/File.php]: OK, changed to `strpos()`.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 77493bd61871aa61a834abf42423807ea42e5bf7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 77493bd61871aa61a834abf42423807ea42e5bf7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by HLeithner** [comment]
The assumption is wrong:
`-rw-r--r--   1 root root    0 Sep 19 17:13 filewith.extentionand\behindit`

This is a valid file name on posix systems...

better way maybe to use `basename()` before extracting the extension.

**Comment by Fedik** [comment]
> This is a valid file name on posix systems

interesting, I did not knew

**Comment by Fedik** [comment]
seems the bot also thinks differently :smile:

**Comment by SharkyKZ** [comment]
Set back to Pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by richard67** [comment]
@SharkyKZ I'm not sure if I like the pathinfo thing. I prefer to see what is expected when reading the test. But maybe that's just my taste. Other changes look good to me.

**Comment by HLeithner** [comment]
The code looks really ugly but ok.

Whats the reason not to use pahinfo for extension extraction? Performance?

And if so please add a comment that this is the reason.

**Comment by SharkyKZ** [comment]
Yes, performance. But whatever maintainers decide. PR https://github.com/joomla/joomla-cms/pull/18030 using `pathinfo()` has been sitting open for a while.

As a compromise between performance, readability and reliability we could use `SplFileInfo::getExtension()`.

**Comment by richard67** [comment]
@SharkyKZ It's one thing if we use it in the CMS code. I would be ok with that. It's another thing if we should use it in the test, too.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 263bfcc9a48e95c45dcad79799a16a3b8449fca6<br /><br />PR does what it claims to do.
I'm not a fan of using pathinfo in the test because before that change, I could see the hard-coded expected result.
But that's maybe just a matter of taste.
I've tested that the expected result is equal to the extension provided by pathinfo for all cases handled by the new test and a few more.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by HLeithner** [comment]
I'm happy with the tests and that you use phpinfo so we know at least that we do the same then php.

Could you just add the comment about performance into the code and get a second test then I will merge it.

**Comment by richard67** [comment]
Last change was bigger so my previous test isn't valid anymore.

@SharkyKZ Do you know where I can see in drone or appveyor if the new tests added by this PR have run?

**Comment by SharkyKZ** [comment]
No idea.

**Comment by richard67** [comment]
@HLeithner Are the tests ok now after the last change?

I don't really understand how it works. It adds a new data provider and a new test routine. But I don't see anything being added where those are called. Does that happen automatically, i.e. the test goes through test PHP files and scans for test routines and corresponding data provides, if any, according to a pre-defined naming scheme to collect which tests to run with which data?

And is there a place in logs here where I can see that they have run? Or do I have to run them in my local environment to verify that?

**Comment by HLeithner** [comment]
@SharkyKZ can you please add the comment why you used this instead of pathinfo to the code (we need much more of such comments) before the next person comes and want to change it to pathinfo

@richard67 we need a second test again maybe from @Fedik ?

The tests are executed automatically based on the name (iirc starting with test*), phpunit executes the data provider and iterates over the returned array and executes the test function with the value.

**Comment by SharkyKZ** [comment]
Added comment.

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on eeb21873db80c2c0d24ac5c4a322a2c2b7f911d3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on eeb21873db80c2c0d24ac5c4a322a2c2b7f911d3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30646">issues.joomla.org/tracker/joomla-cms/30646</a>.</sub>

**Comment by richard67** [comment]
@HLeithner Ready for merge.

**Comment by HLeithner** [comment]
Thanks

**Pull request merged**